File: app/models/role.rb
=begin

Copyright (c) 2020, all rights reserved.

All the information provided by this platform is protected by international laws related  to
industrial property, intellectual property, copyright and relative international laws.
All intellectual or industrial property rights of the code, texts, trade mark, design,
pictures and any other information belongs to the owner of this platform.

Without the written permission of the owner, any replication, modification,
transmission, publication is strictly forbidden.

For more information read the license file including with this software.

// · ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~
// ·

=end

class Role < ApplicationLesliRecord

    belongs_to :account,   foreign_key: "accounts_id"


    # Role resources
    has_many :activities,  foreign_key: "roles_id"
    has_many :descriptors, foreign_key: "roles_id", dependent: :delete_all
    has_many :privileges,  foreign_key: "roles_id", class_name: "Role::Privilege", dependent: :delete_all
    

    # initializers for new roles
    before_create :before_create_role
    after_create :after_create_role, :initialize_role_privileges


    # validations
    validates :name, presence: :true
    validates :object_level_permission, presence: :true


    def before_create_role

        # default path for limited roles
        if self.name == "limited"
           self.path_default ||= "/administration/profile" 
        end

        # enable roles by default
        self.active = true

    end

    def after_create_role

        # generate a unique code for the role
        role_code = name
            .downcase                           # string to lowercase
            .gsub(/[^0-9A-Za-z\s\-\_]/, '')     # remove special characters from string
            .gsub(/-/, '_')                     # replace dashes with underscore
            .gsub(/\s+/, '_')                   # replace spaces or spaces with single dash

        # transform UTF-8 characters to ASCI (ex: name in german)
        role_code = I18n.transliterate(role_code) + id.to_s 

        self.update_attribute("code", role_code)
    end

    # default roles must have privileges by default
    def initialize_role_privileges
        return 
        if (self.name == "owner" || self.name == "sysadmin" || self.name == "limited")
            descriptor = self.name
            descriptor = "profile" if descriptor == "limited"
            SystemDescriptor.find_or_create_by(
                descriptor: SystemDescriptor.find_by(name: descriptor)
            )
        end
    end

    # @return [Boolean]
    # @description Returns if a role is assigned to users.
    def has_users?
        User::Role.where(role: self).count > 0
    end

    # @return [void]
    # @param current_user [::User] The user that deleted the role
    # @param [Role] The role that was deleted
    # @description Creates an activity for this role indicating that someone deleted it
    # Example
    #   role = Role.find(1)
    #   Role.log_activity_destroy(User.find(1), role)
    def self.log_activity_destroy(current_user, role)
        role.activities.create(
            user_creator: current_user,
            category: "action_destroy"
        )
    end
end

File: app/vue3/administration/stores/role.js
/*
Copyright (c) 2023, all rights reserved.

All the information provided by this platform is protected by international laws related  to 
industrial property, intellectual property, copyright and relative international laws. 
All intellectual or industrial property rights of the code, texts, trade mark, design, 
pictures and any other information belongs to the owner of this platform.

Without the written permission of the owner, any replication, modification,
transmission, publication is strictly forbidden.

For more information read the license file including with this software.

// · ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~
// · 
*/


// · 
import { defineStore } from "pinia"


// · 
export const useRole = defineStore("administration.role", {
    state: () => {
        return {
            options: {},
            records: [],
            descriptors: [],
            descriptorsCustom: [],
            descriptorSearchString: '',
            role: {
                name: "",
                descriptors: []
            },
            pagination: {
                page: 1
            },
            index: { 
                pagination: {},
                records: []
            },
            loading: false,             
            order: {
                column: "id",
                direction: "desc"
            },
            search_string: ""
        }
    },
    actions: {

        fetch() {
            this.loading = true
            this.http.get(this.url.admin("roles")
                .paginate(this.pagination.page)
                .order(this.order.column, this.order.direction)
                .search(this.search_string))
            .then(result => {
                this.index = result
                this.records = result.records.map(integrations => {
                    integrations.created_at = this.date.dateTime(integrations.created_at)
                    return integrations
                })
                this.loading = false
            })
        },

        paginateIndex(page) {
            this.pagination.page = page
            this.fetch()
        },

        fetchRole(id) {
            this.loading = true
            this.http.get(this.url.admin("roles/:id", id)).then(result => {
                this.role = result
                this.getDescriptors()
                this.loading = false
            })
        },

        postRole() {
            return this.http.post(this.url.admin("roles"), this.role)
        },

        putRole() {
            this.http.put(this.url.admin("roles/:id", this.role.id), this.role).then(result => {
                this.msg.success(I18n.t('core.roles.messages_success_role_successfully_updated'))
            })
        },

        searchDescriptors(string) {
            this.descriptorSearchString = string
            this.getDescriptors()
        },

        getDescriptors() {

            let url = this.url.admin("roles/:id/describers", this.role.id)

            if (this.descriptorSearchString != '') {
                url = url.search(this.descriptorSearchString)
            }

            this.http.get(url).then(descriptors => {

                // reset the list of descriptors
                this.descriptors = []
                this.descriptorsCustom = []

                // temporary container for the descriptor matrix
                let descriptorgrid = {}

                // convert list of descriptors from: controller/action to controller/actions (like a matrix)
                descriptors.forEach(descriptor => {

                    if (['index', 'show', 'new', 'edit', 'destroy', 'search'].includes(descriptor.action)) {

                        if (!descriptorgrid[descriptor.reference]) {
                            descriptorgrid[descriptor.reference] = {
                                id: descriptor.id,
                                name: descriptor.name,
                                reference: descriptor.reference,
                                controller: descriptor.controller,
                                index: null, 
                                show: null, 
                                create: null, 
                                update: null, 
                                destroy: null
                            }
                        }

                        // add the id of the descriptor that the action belongs to
                        descriptorgrid[descriptor.reference][descriptor.action] = {
                            id: descriptor.id,
                            active: descriptor.active
                        }

                        return;

                    }

                    this.descriptorsCustom.push(descriptor)

                })

                // return the arrys only
                this.descriptors = Object.values(descriptorgrid)

            })
        },

        updateDescriptor(descriptor) {

            // enable descriptor
            if (descriptor.active) {
                return this.postDescriptor(descriptor)
            }

            // disable descriptor
            this.deleteDescriptor(descriptor)
            
        },

        postDescriptor(descriptor) {
            this.http.post(this.url.admin("roles/:id/describers", this.role.id), {
                role_describer: {
                    id: descriptor.id
                }
            })
        },

        deleteDescriptor(descriptor) {
            this.http.delete(this.url.admin("roles/:id/describers/:descriptor_id", {
                id: this.role.id,
                descriptor_id: descriptor.id
            }))
        },

        getOptions() {
            this.http.get(this.url.admin("roles/options")).then(result => {
                
                this.options = result
            })
        }, 
        /**
         * @description This action is used to sort the results
         */
        sortIndex(column, direction) {
            this.order.column = column
            this.order.direction = direction
            this.fetch()
        },
        /**
         * @description This action is used to search
         * @param {string} search_string 
         */
        search(search_string) {
            this.search_string = search_string
            this.fetch()
        }

    }
})
